### PR TITLE
Fix issues with Suspend/Resume in BootStrapper

### LINKED ISSUE #1386: Rare crash in Bootstrapper.SetupLifecycleListeners  on suspend
Sometimes we get exceptions on suspend at this code in then Suspending action in Bootstrapper.SetupLifecycleListeners:

```
// application-level
var isPrelaunch = (OriginalActivatedArgs as LaunchActivatedEventArgs)?.PrelaunchActivated ?? false;
await OnSuspendingAsync(this, e, isPrelaunch);
```

Problem is caused because the OriginalActivatedArgs are a system object (COM based) and got invalid. So the app crashes when it tries to access this object on suspend. The Crash itself looks like this:

```
Thread 2 Crashed:
0   XXX.Uap10                        0x00007fffcfd9927d Template10.Utils.Template10Utils.GetDispatcherWrapper($INavigationService service) in E:\Build1\71cdfcbed6e20846\3rdParty\Template10\Template10 (Library)\Utils\Template10Utils.cs at 30:16
1   XXX.Uap10                        0x00007fffd03c42a1 XXX.Uap10.App.<OnResumingAsync>d__7.MoveNext() in E:\Build1\71cdfcbed6e20846\XXX\XXX.Uap10\App.xaml.cs at 280:4
2   XXX.Uap10                        0x00007fffcfbb2531 System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw() in f:\dd\ndp\fxcore\CoreRT\src\System.Private.CoreLib\src\System\Runtime\ExceptionServices\ExceptionDispatchInfo.cs at 61:13
3   XXX.Uap10                        0x00007fffcfbb442e System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess($Task task) in f:\dd\ndp\fxcore\CoreRT\src\System.Private.Threading\src\System\Runtime\CompilerServices\TaskAwaiter.cs at 187:25
4   XXX.Uap10                        0x00007fffcfbb43a8 System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification($Task task) in f:\dd\ndp\fxcore\CoreRT\src\System.Private.Threading\src\System\Runtime\CompilerServices\TaskAwaiter.cs at 156:42
5   XXX.Uap10                        0x00007fffcfbbcf1a System.Runtime.CompilerServices.ConfiguredTaskAwaitable.ConfiguredTaskAwaiter.GetResult() in f:\dd\ndp\fxcore\CoreRT\src\System.Private.Threading\src\System\Runtime\CompilerServices\TaskAwaiter.cs at 113:13
6   XXX.Uap10                        0x00007fffd031a6d7 Neusta.Commons.Template10.Utils.ExtendedBootStrapper.<OnResuming>d__11.MoveNext() in E:\Build1\71cdfcbed6e20846\Neusta\Neusta.Commons.Template10.Universal\Utils\ExtendedBootStrapper.cs at 16707566:0
7   XXX.Uap10                        0x00007fffcfbb2531 System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw() in f:\dd\ndp\fxcore\CoreRT\src\System.Private.CoreLib\src\System\Runtime\ExceptionServices\ExceptionDispatchInfo.cs at 61:13
8   XXX.Uap10                        0x00007fffcfbbf40e System.Runtime.CompilerServices.AsyncMethodBuilderCore.<>c.<ThrowAsync>b__9_0(Object state) in f:\dd\ndp\fxcore\CoreRT\src\System.Private.Threading\src\System\Runtime\CompilerServices\AsyncMethodBuilder.cs at 957:49
9   XXX.Uap10                        0x00007fffcfd7a84c System.Threading.WinRTSynchronizationContext.Invoker.InvokeCore() in f:\dd\ndp\fxcore\CoreRT\src\System.Private.Threading\src\System\Threading\SynchronizationContext.cs at 171:21
10  XXX.Uap10                        0x00007fffcfba22d0 McgInterop.McgHelpers.ThrowFailed(Int32 hr, RuntimeTypeHandle typeHnd) in c:\windows\ServiceProfiles\NetworkService\appdata\local\packages\appcompile-20170116-061302-12832\ac\Temp\Ilc52466529\XXX.Uap10.McgInterop\Helpers.g.cs at 6416:10
11  XXX.Uap10                        0x00007fffcfc385e5 McgInterop.ComCallHelpers.ComCall__HRESULT($__ComObject __this, RuntimeTypeHandle __typeHnd, Int32 __targetIndex) in c:\windows\ServiceProfiles\NetworkService\appdata\local\packages\appcompile-20170116-061302-12832\ac\Temp\Ilc52466529\XXX.Uap10.McgInterop\SharedStubs.g.cs at 271:5
12  XXX.Uap10                        0x00007fffcfc3850b McgInterop.ForwardComSharedStubs.Proc_<System.__Canon>(Void* InstParam, $__ComObject __this, Int32 __targetIndex) in c:\windows\ServiceProfiles\NetworkService\appdata\local\packages\appcompile-20170116-061302-12832\ac\Temp\Ilc52466529\XXX.Uap10.McgInterop\SharedStubs.g.cs at 4983:4
13  XXX.Uap10                        0x00007fffd008e0e8 Windows.ApplicationModel.Core.UnhandledError.Propagate() in c:\windows\ServiceProfiles\NetworkService\appdata\local\packages\appcompile-20170116-061302-12832\ac\Temp\Ilc52466529\XXX.Uap10.McgInterop\SafeTypes.g.cs at 5849:4
14  XXX.Uap10                        0x00007fffd008b41d Microsoft.HockeyApp.Extensibility.Windows.UnhandledExceptionTelemetryModule.CoreApplication_UnhandledErrorDetected(Object sender, $UnhandledErrorDetectedEventArgs e) in D:\Projekte\3rdParty\HockeySDK-Windows\Src\Kit.UWP\Services\UnhandledExceptionTelemetryModule.cs at 201:17
```

### PR DESCRIPTION
**Pull request by schuettecarsten**
See commit messages, this PR should fix #1386 and #1387.
Please review carefully for unexpected side effects.

### CODE REVIEW

**Comment by schuettecarsten** [comment]
@Robzilla: Do you mean OnNavigatedFrom or OnNavigatingFrom? This behaviour is unrelated to my changes and seems to be "by design" for Template10. You have to check the "suspending" parameter on OnNavigatingFromAsync.

@JerryNixon: What do you think about adding OnResumeAsync to base viewmodel?

**Comment by tibitoth** [comment]
Call OnNavigatedToAsync IMO

**Comment by schuettecarsten** [comment]
@totht91: But this will be a breaking change. Please also keep in mind that OnNavigatingFrom is some kind of "Before"-event that allows to cancel navigation. If you need to unload data, you should use OnNavigatedFrom.

**Comment by JerryNixon** [comment]
It is implied in OnNavigatingFrom that the event can be canceled and prevent continuation. In the case of Suspension, there is no such opportunity. That being said, this is not obvious to the developer. Even so, a developer using OnNavigatingFrom (the before event) to do important work INSTEAD of OnNavigatedFrom  (the after event) has already strayed from the guidance in the templates. So I think this is okay, we can assume this developer that (for whatever reason) has decided to use OnNavigatingFrom for purposes other than preventing navigation is advanced and can deal with this subtle implementation. (Which represents the current impl.) I say all that to say this, I think OnNavigatingFrom should NOT be called during SUSPEND.  If for no other reason, because the cancel `eventarg` property is misleading. After all, OnNavigatedFrom already has a `suspending` argument, and the cancelable implication of OnNavigatingTo makes it a bad match. @schuettecarsten, I agree with you here. For sure. I really did reconsider for a second.

File: Template10 (Library)/Common/Bootstrapper/BootStrapper.cs [hunk 1]
```diff
@@ -71,34 +71,46 @@ public enum AdditionalKinds { Primary, Toast, SecondaryTile, Other, JumpListItem
...
         /// </summary>
         public bool AutoSuspendAllFrames { get; set; } = true;
 
         /// <summary>
         /// There are many ways for an app to start and re-start, including activation.
-        /// This property retains the original (first-launch) activation arguments. 
+        /// This field retains the original (first-launch) activation arguments.
         /// </summary>
-        public IActivatedEventArgs OriginalActivatedArgs { get; private set; }
+        private IActivatedEventArgs _originalActivatedArgs;
```
Comment (JerryNixon) [inline Template10 (Library)/Common/Bootstrapper/BootStrapper.cs:101]: Please explain why it is important this not be accessible externally.

File: Template10 (Library)/Common/Bootstrapper/BootStrapper.cs [hunk 2]
```diff
@@ -345,15 +357,15 @@ public BootstrapperStates CurrentState
             get { return _currentState; }
             set
             {
-                DebugWrite($"CurrenstState changed to {value}");
+                DebugWrite($"CurrentState changed to {value}");
```
Comment (JerryNixon) [inline Template10 (Library)/Common/Bootstrapper/BootStrapper.cs:360]: Thank you. Trivial doesn't mean unimportant.

File: Template10 (Library)/Common/Bootstrapper/BootStrapper.cs [hunk 3]
```diff
@@ -71,34 +71,46 @@ public enum AdditionalKinds { Primary, Toast, SecondaryTile, Other, JumpListItem
...
+        /// There are many ways for an app to start and re-start, including activation.
+        /// This property retains the previous execution state.
+        /// </summary>
+        public ApplicationExecutionState PreviousExecutionState { get; private set; }
+
+        /// <summary>
+        /// There are many ways for an app to start and re-start, including activation.
+        /// This property retains a fklag indicating whether Prelaunch is activated.
+        /// </summary>
+        public bool PrelaunchActivated { get; private set; }
```
Comment (JerryNixon) [inline Template10 (Library)/Common/Bootstrapper/BootStrapper.cs:113]: Does this mean the current activation is prelaunch or just that there was one sometime?

File: Template10 (Library)/Common/Bootstrapper/BootStrapper.cs (hunk 1 above)
Comment (schuettecarsten) [reply Template10 (Library)/Common/Bootstrapper/BootStrapper.cs:101]: The _originalActivatedArgs is useless after my changes. It is only used to check if the current activation is the first one. If can be replaced by a bool value.

File: Template10 (Library)/Common/Bootstrapper/BootStrapper.cs (hunk 3 above)
Comment (schuettecarsten) [reply Template10 (Library)/Common/Bootstrapper/BootStrapper.cs:113]: It means that if was PrelaunchActivated on the *first* launch, at the moment when _originalActivatedArgs was set.

**Pull request merged**